src/osp/Active/SysNewton.cpp

```diff
@@ -486,18 +492,34 @@ void SysNewton::body_apply_torque(ACompRigidBody_t &body, Vector3 torque) noexce
     body.m_netTorque += torque;
 }
 
-Vector4 SysNewton::compute_body_CoM(ActiveScene& rScene,
-    ActiveEnt root, Matrix4 currentTransform)
+/* Since masses are usually stored in the rigidbody's children instead of the
+ * rigidbody root, the function provides the option to ignore the root entity's
+ * mass (this is the default). Otherwise, the root of each call will be double
+ * counted since it will also be included in both the loop and the root of the
+ * next call. This makes it possible for a rigidbody to have a convenient
+ * root-level mass component set by some external system without interfering
+ * with physics calculations that depend on mass components to determine the
+ * mechanical behaviors of the rigidbody.
+ */
+Vector4 SysNewton::compute_hier_CoM(ActiveScene& rScene,
+    ActiveEnt root, bool includeRootMass)
 {
     Vector3 localCoM{0.0f};
     float localMass = 0.0f;
 
-    for (ActiveEnt nextChild = root; nextChild != entt::null;)
+    if (includeRootMass)
     {
-        auto const &childHeir = rScene.reg_get<ACompHierarchy>(nextChild);
+        auto const* rootMass = rScene.reg_try_get<ACompMass>(root);
+        if (rootMass != nullptr) { localMass += rootMass->m_mass; }
```

Comment:
tying back to our discussion about exceptions, i'm wondering about the benefit of having an ActiveEnt that represents a root that can't be located?

That seems like an "exceptional" situation that we should make the game bork-out about, instead of just adding nothing to the mass.

Reply:
In this case, the thing being found is the mass component, so `rootMass` is type `ACompMass`. `try_get<T>(ent)` tries to find component `T` belonging to entity `ent`, and if `ent` doesn't have a `T` then nullptr is returned. In this case, `includeRootMass` simply indicates that you're interested in the root node's mass if it has one, and if it doesn't you don't care (it skips it just like this function skips any other massless entity).

Reply:
That's fair